### PR TITLE
Improve detecting permissions to view referenced discussions

### PR DESCRIPTION
**Pull request by rob006**
Replaces #30

This PR fixes detecting permissions for discussions visibility, as `$actor->can('viewForum', $discussion)` sytnax is probably [incorrect](https://discord.com/channels/360670804914208769/360689688815730701/1076217045475143720) and did not worked for me.

It also improves behavior in case of lack of permissions or errors caused by other extensions:
1. Fallbacks user to Guest in case if request or actor was not passed to formatter - it should fix all issues for public forums.
2. Instead displaying "unknown discussion" placeholder, it leaves URL intact, so it is still clickable.
3. logs errors only when debug mode is enabled.

### CODE REVIEW

File: js/src/forum/index.ts [hunk 1]
```diff
@@ -54,12 +53,7 @@ function addSourceLinkReplacement() {
       if (match == null) {
         return;
       }
-      if (a.text === a.href) {
-        const discussionId = match[1];
-        const span = document.createElement('span');
-        m.mount(span, {view: () => m(DiscussionLink, {discussionId, href: a.href})});
-        a.replaceWith(span)
-      } else {
```
Comment (n-peugnet) [inline js/src/forum/index.ts:62]: I want to keep this. It is used to retrofit old comments posted before the extension was enabled.

File: src/Formatter/CrossReferencesConfigurator.php [hunk 2]
```diff
@@ -71,21 +72,22 @@ public function __invoke(Configurator $config): void
     public static function filterCrossReferences(Tag $tag, ?User $actor): bool
     {
         if (is_null($actor)) {
-            $log = resolve(LogReporter::class);
-            $msg = 'actor is "null", falling back to invalidating tag. This is probably due to another extension not passing this parameter to "Formatter->parse()". See stack trace below.';
-            $log->report(new RuntimeException($msg));
-            $tag->invalidate();
-            return false;
+            if (resolve('flarum.config')->inDebugMode()) {
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesConfigurator.php:75]: Why only log the error in development mode? How are supposed the admins to understand the issue if no error is logged? And what is the benefit of not logging it?

File: src/Formatter/CrossReferencesConfigurator.php [hunk 3]
```diff
@@ -71,21 +72,22 @@ public function __invoke(Configurator $config): void
...
+                $log = resolve(LogReporter::class);
+                $msg = 'actor is "null", falling back to display discussion as for Guest. This is probably due to another extension not passing this parameter to "Formatter->parse()". See stack trace below.';
+                $log->report(new RuntimeException($msg));
+            }
+            $actor = new Guest();
         }
         /** @var Discussion|null */
-        $d = Discussion::find($tag->getAttribute('id'));
-        if (is_null($d) || $actor->cannot('viewForum', $d)) {
+        $discussion = Discussion::whereVisibleTo($actor)->firstWhere('id', $tag->getAttribute('id'));
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesConfigurator.php:83]: The mocks in unit tests will have to be updated.

File: src/Formatter/CrossReferencesConfigurator.php [hunk 4]
```diff
@@ -100,11 +102,9 @@ protected function configureCrossReferenceShort(Configurator $config): void
         $tag->template = '
             <xsl:choose>
                 <xsl:when test="@unknown = 1">
-                    <span class="DiscussionLink DiscussionUnknown">
+                    <a href="{$DISCUSSION_URL}{@id}" class="DiscussionLink">
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesConfigurator.php:105]: I don't want to add the link if the discussion is unknown. This will lead to bad experiences from the user getting 404 errors (Flarum does not differentiate between unauthorised and not found). And it is overall not good to have internal links leading to 404 errors.

File: src/Formatter/CrossReferencesConfigurator.php [hunk 5]
```diff
@@ -133,11 +133,9 @@ protected function configureCrossReferenceURL(Configurator $config): void
         $tag->template = '
             <xsl:choose>
                 <xsl:when test="@unknown = 1">
-                    <span class="DiscussionLink DiscussionUnknown">
+                    <a href="{@url}" class="DiscussionLink">
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesConfigurator.php:136]: Same as above.

File: src/Formatter/CrossReferencesConfigurator.php [hunk 6]
```diff
@@ -167,11 +165,9 @@ protected function configureCrossReferenceURLComment(Configurator $config): void
...
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesConfigurator.php:168]: Same as above.

File: src/Formatter/CrossReferencesRenderer.php [hunk 7]
```diff
@@ -60,20 +68,22 @@ public function __construct(Translator $translator, LogReporter $log)
      */
     public function __invoke(Renderer $renderer, $context, ?string $xml, ?ServerRequestInterface $request)
     {
-        $actor = null;
         if (is_null($request)) {
-            $msg = 'request is "null", falling back to display discussions as unknown. This is probably due to another extension not passing this parameter to "Formatter->render()". See stack trace below.';
-            $this->log->report(new RuntimeException($msg));
+            if ($this->config->inDebugMode()) {
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesRenderer.php:72]: Same as above? Why only in dev mode?

File: src/Formatter/CrossReferencesRenderer.php [hunk 8]
```diff
@@ -60,20 +68,22 @@ public function __construct(Translator $translator, LogReporter $log)
...
         $filterCrossReferences = function ($attributes) use ($actor) {
             /** @var Discussion|null */
-            $discussion = Discussion::find($attributes['id']);
-            if ($discussion && $actor && $actor->can('viewForum', $discussion)) {
+            $discussion = Discussion::whereVisibleTo($actor)->firstWhere('id', $attributes['id']);
+            if ($discussion) {
                 $attributes['title'] = $discussion->title;
             } else {
-                $attributes['title'] = $this->translator->trans('club-1-cross-references.forum.unknown_discussion');
+                $attributes['title'] = $attributes['url'];
```
Comment (n-peugnet) [inline src/Formatter/CrossReferencesRenderer.php:86]: I don't think displaying the URL if the discussion is unknown is a good idea.

**Review by n-peugnet — COMMENTED** [review]
> 1.    Fallbacks user to Guest in case if request or actor was not passed to formatter - it should fix all issues for public forums.
> 2.    Instead displaying "unknown discussion" placeholder, it leaves URL intact, so it is still clickable.
> 3.    logs errors only when debug mode is enabled.


I am OK with the first point but I don't think I like the two after.
I don't see the point of keeping a clickable URL leading to a 404 and I made this not an URL on purpose to avoid this.
And I don't understand why we should not log the error in production mode.

> This PR fixes detecting permissions for discussions visibility, as $actor->can('viewForum', $discussion) sytnax is probably [incorrect](https://discord.com/channels/360670804914208769/360689688815730701/1076217045475143720) and did not worked for me.

I can't access the content of this discord link (I refuse to create an account on this service), but you are probably right about this. It was difficult to find information about the permission system of Flarum, so I mostly did it by guessing.

File: src/Formatter/CrossReferencesConfigurator.php (hunk 4 above)
Comment (rob006) [reply src/Formatter/CrossReferencesConfigurator.php:105]: But you don't know if it is 404 - permissions detection may be incorrect and link could actually work. In this way extension improves behavior for safe cases and do not break anything in uncertain scenarios.

File: src/Formatter/CrossReferencesConfigurator.php (hunk 2 above)
Comment (rob006) [reply src/Formatter/CrossReferencesConfigurator.php:75]: In many cases this log message is irrelevant. In my case forum is public, it does not matter if extension can't detect actor, because fallback to guest will handle this correctly in 99%. and for the rest 1% displaying raw link is fine. I don't need a logs full of these messages, because I do not have plans to fix extensions that may cause it. And if someone wants to understand why links are not parsed, debug mode is the right way to do this.

File: js/src/forum/index.ts (hunk 1 above)
Comment (rob006) [reply js/src/forum/index.ts:62]: Isn't that super inefficient, because each discussion will create separate request to API? So If I have 100 links in my post, it will trigger 100 requests to API?

Comment (n-peugnet) [reply js/src/forum/index.ts:62]: True if you have 100 links, each pointing to different discussions. But I would say this is rather uncommon. I think a better way to solve this problem would be to combine these multiple requests into one in the [ResponseCache](https://github.com/club-1/flarum-ext-cross-references/blob/7e602c22cc4c799b7a8c07afba5b4da16268424e/js/src/forum/cache.ts#L83). I could do it if this is a real problem for you.

Anyways I would prefer to do this in a separate pull request.

File: src/Formatter/CrossReferencesConfigurator.php (hunk 2 above)
Comment (n-peugnet) [reply src/Formatter/CrossReferencesConfigurator.php:75]: OK I guess, It is true that it adds a lot of noise in the logs. Maybe only logging one line in production mode instead of the full stack trace would be a good compromise? While keeping the stack trace in debug mode?

File: src/Formatter/CrossReferencesConfigurator.php (hunk 4 above)
Comment (n-peugnet) [reply src/Formatter/CrossReferencesConfigurator.php:105]: Isn't the Guest fallback enough for most use cases? And isn't [this change](https://github.com/club-1/flarum-ext-cross-references/pull/31/files#diff-18af1e3972585e1b07ba509ff590f4281ac1232144731d14c87a4cd5a0f74e85R82) enough to fix the detection in your case? For instance would this patch be enough: https://github.com/club-1/flarum-ext-cross-references/commit/fae283253f3dee629214088b2eed2302ba9c087b

I don't mind marking this extension incompatible with others for some time while the render/parse calls are incomplete. For instance `flarum/mentions` just crashes in these cases.

**Comment by rob006** [comment]
I think that there are different expectations from this extension, so I share my POV.

I installed this extension for backlink feature - it is really convenient in a forum used for gathering feedback, where sooner or later you get a bunch of duplicated discussions, which you want to link to the canonical one. Converting original links, so they display a discussion title, is a nice bonus, but right now it has way too many drawbacks:

1. **Performance**. Even server-side parsing is sub-optimal, since it queries discussion one by one, so 100 links = 100 extra queries. But that is acceptable, it should not happen that often. The same problem on JS level is way worse, but I can live with that too. But the fact that there is a visible delay between loading the original link and replacing it with the discussion title is too much - it shifts post content and it looks really ugly. I really think that applying this feature to old posts should be done by re-parsing old posts content by some console command and then relying on standard server-side logic, instead of having a worse substitute on JS level.

2. **Placeholder for unavailable discussions**. Replacing the original link with an unclickable placeholder is no-go in my case. Even if we ignore the possibility, that detection may be incorrect, you're just hiding useful information, like ID or slug, that may explain the intention of this link. There is also compatibility with other extensions or tools - some of them may operate on different permissions than consumer of output. For example, an external tool that converts discussion to PDF may not have permission to see a particular discussion, but a user that will read this PDF may have such permission, and hiding these links from him is a regression. Or extension that handles redirections - AFAIK it is not implemented yet, but it makes perfect sense to handle redirection if 2 discussions were merged, so deleted one will redirect to a target discussion. Right now this extension hides links, that may actually work.
   IMO this feature should be optional, and I strongly advise disabling it by default. 

3. **Logging**.  The inability to completely disable these logs for production is IMO unacceptable. Even skipping stack trace is not a solution, because this may still result in thousands of entries in the error log. I think that you're overestimating both the importance and usefulness of these logs, and it is mostly a result of an uncompromising approach for handling placeholders for unavailable discussion - right now bad things may happen if we have incompatible extensions because links no longer work. But it does not have to be like that - if we leave links intact, then nothing is broken, and it is safe to ignore this issue as long as the output is acceptable. And it should be acceptable in most cases since this is default Flarum behavior. ;)
   If you want to make it easier to debug this on production environment, then having a separate switch in settings to disable these logs is the way to go. Multiple extensions have something like that.

So to summarize my proposal:
1. Logging "errors" should be switchable in extension settings and disabled by default.
2. Replacing links to potentially inaccessible discussions with placeholder should be switchable in extension settings and disabled by default. This applies only to links in post content - current behavior for `DiscussionReferencedPost` is correct.
3. Fixing old posts should be done by re-parsing content using console command.

**Comment by rob006** [comment]
I have no idea how to fix `Static method Flarum\Discussion\Discussion::whereVisibleTo() does not exist on this mock object` error. I could try adjust the rest of the tests if we agree on desired behavior, but mocking this Laravel magic (or configuring real tests with database involved) is beyond me.

**Comment by n-peugnet** [comment]
> I have no idea how to fix Static method Flarum\Discussion\Discussion::whereVisibleTo() does not exist on this mock object error. I could try adjust the rest of the tests if we agree on desired behavior, but mocking this Laravel magic (or configuring real tests with database involved) is beyond me.

I will do it. This is indeed tricky and I might also have pushed the limits of unit testing too far :sweat_smile:



> So to summarize my proposal:
> 
> 1. Logging "errors" should be switchable in extension settings and disabled by default.
> 
> 2. Replacing links to potentially inaccessible discussions with placeholder should be switchable in extension settings and disabled by default. This applies only to links in post content - current behavior for `DiscussionReferencedPost` is correct.
> 
> 3. Fixing old posts should be done by re-parsing content using console command.


Thank you for taking the time to write these detailed explanations, I get your points.

1. OK for logging only in debug mode. I could add an FAQ about why the rendering is incorrect that points to enabling debug mode.
2. You are right, removing the link may not be as important in posts' content as in the `DiscussionReferencedPost`. Adding an option for this is fine for me.
3. This can be added later. If you have some time you can open an issue about this, or else I will do it myself.

I will extract the "Guest changes" and point 1. from your pull request and merge then right now (started here: https://github.com/club-1/flarum-ext-cross-references/pull/32). And I will see what I can do for 2. that would be acceptable for you.

About this point specifically:

> Converting original links, so they display a discussion title, is a nice bonus, but right now it has way too many drawbacks:
> 
> 1. **Performance**. Even server-side parsing is sub-optimal, since it queries discussion one by one, so 100 links = 100 extra queries. But that is acceptable, it should not happen that often.

You are right, fetching all discussions in one request while rendering the posts in the backend would be a nice performance enhancement, and not that difficult to do.

**Comment by n-peugnet** [comment]
@rob006 I did this part of the changes in #32:

> I will extract the "Guest changes" and point 1. from your pull request and merge them right now

If this is good for you, I will merge it now. I will then work on your second point and come back to you later today or tomorrow.

File: js/src/forum/index.ts (hunk 1 above)
Comment (n-peugnet) [reply js/src/forum/index.ts:62]: > I think a better way to solve this problem would be to combine these multiple requests into one in the [ResponseCache](https://github.com/club-1/flarum-ext-cross-references/blob/7e602c22cc4c799b7a8c07afba5b4da16268424e/js/src/forum/cache.ts#L83). I could do it if this is a real problem for you.


FWIW I implemented a proof of concept of this idea in https://github.com/club-1/flarum-ext-cross-references/commit/1d79290a766b98c515256345591dc26bc96e415e
But it is currently blocked by https://github.com/flarum/framework/issues/3739

Comment (rob006) [reply js/src/forum/index.ts:62]: I think that there should be a way to include data of referenced discussions in original payload. Then you don't need any additional requests or custom memoization, you can use default Flarum storage to fetch discussion data.

Comment (n-peugnet) [reply js/src/forum/index.ts:62]: > I think that there should be a way to include data of referenced discussions in original payload.

This was the way it was done at first when I was not taking permissions into account. But this is indeed an improvement that I want to make and is tracked here: https://github.com/club-1/flarum-ext-cross-references/issues/13

But this will not work for the "retrofit in JS" use case as these relations are not stored in the database yet (it would be fixed by the re-parsing of messages that you suggested).
